### ISSUE TITLE
Blender / GLTF: Add Scene.environment intensity

### ISSUE THREAD

**Issue by mjurczyk** [mrdoob/three.js#24665]
Playing around with [Blender HDRI workflows](https://www.youtube.com/watch?v=z8IGDM2f2SQ) - and it looks seems there's no reasonable way of lowering the environment map intensity, to use it only as a supporting light (not that it's a vital feature - but a tiny missing link between Three and Blendor / [glTF IBLs ref](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Vendor/EXT_lights_image_based/README.md#image-based-light-properties).)

<img width="210" alt="kot" src="https://user-images.githubusercontent.com/9549760/190911343-ba2af6f7-a83f-4141-ab6f-cd4d871523f8.png">

Since `scene.environment` is a "global" setting - would it make sense to add [scene.environmentIntensity](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L1460) to `uniforms` - overriding `.envMapIntensity` if it is not defined? A bit ugly, but it'd be following the way [envMap is currently inherited](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L1376) from the Scene.

(A cleaner way may probably be to split `scene.environment` as ` = { map: [IBLTexture], intensity: [Number] }` and add backwards-compatibility in the setter?) 

(A way cleaner way would be probably to add `.intensity` directly as a new property of a Texture to generalise it - since there's already quite a few intensity-based textures in PBR materials - but that'll probably not be super backwards-compatible?)

**Comment by LeviPesin** [comment]
Can't we use the A channel for intensity? Opacity and intensity are quite related things for textures.

**Comment by Mugen87** [comment]
> Since scene.environment is a "global" setting - would it make sense to add [scene.environmentIntensity](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L1460) to uniforms - overriding .envMapIntensity if it is not defined?

I think it's better for asking users to change the `envMapIntensity` of the material instead of introducing a texture property that overwrites it. Otherwise devs could request the same for all other intensity types (`aoMapIntensity`, `lightMapIntensity`) which only introduces an unnecessary complexity.

**Comment by viridia** [comment]
In the game I am working on, it's not really feasible to adjust the `envMapIntensity` of individual materials. I have materials coming from many different sources - some loaded from GLTF models, some are created programmatically. Some materials are owned by the terrain subsystem, the scenery subsystem, the actor subsystem and so on, and in many cases the reference to a given material may be buried deep within a complex data structure. Creating an iterator that is able to visit all materials, while remaining leak-free (that is, not holding on to references to materials after they are no longer needed) would be a challenge.

An easy way to make a global `envMapIntensity` backwards compatible would be to set the default global intensity to 1, and then make the actual intensity the product of the global setting and the per-material setting. This avoids the need to check for null / undefined during processing, it's just a multiply.

**Comment by viridia** [comment]
Also, I think the slippery-slope argument doesn't hold, because there isn't a global aoMap setting. The `environment` property is special because it's a property of the scene - other map types don't have a global settings (other than background).

**Comment by mrdoob** [comment]
Now that we have `backgroundIntensity` (https://github.com/mrdoob/three.js/pull/24876) maybe it does make sense to add `environmentIntensity` too... 

/cc @WestLangley

**Comment by WestLangley** [comment]
I can add that if it is desired.

Thing is, regarding `scene.environment`, the docs say:

>However, it's not possible to overwrite an existing texture assigned to MeshStandardMaterial.envMap.

So, in that regard, would is be best to have `scene.environmentIntensity` apply only to `scene.environment`, and `mesh.envMapIntensity` apply only to `mesh.envMap`?

**Comment by mrdoob** [comment]
@WestLangley 

> So, in that regard, would it be best to have `scene.environmentIntensity` apply only to `scene.environment`, and `mesh.envMapIntensity` apply only to `mesh.envMap`?

That makes sense to me.

**Comment by drcmda** [comment]
> I can add that if it is desired.
> 
> Thing is, regarding `scene.environment`, the docs say:
> 
> > However, it's not possible to overwrite an existing texture assigned to MeshStandardMaterial.envMap.
> 
> So, in that regard, would it be best to have `scene.environmentIntensity` apply only to `scene.environment`, and `mesh.envMapIntensity` apply only to `mesh.envMap`?

i think this is too big of a breaking change, but also limiting. it should probably regulate env map intensity, whichever env map is used in the end, either the global or the local.

**Comment by mrdoob** [comment]
> personally, i have hundreds of sandboxes where i use scene.environment + individual mesh.envMapIntensity.

Oh...

So `scene.environmentIntensity * material.envMapIntensity` then?
That should be backwards compatible and also like `material.color` and `material.map` work together.

**Comment by WestLangley** [comment]
>scene.environmentIntensity * material.envMapIntensity then?

After trying to implement API, it seems weird to me...

The environment map = `scene.environment` OR `mesh.envMap` -- one or the other.

Yet, the intensity is the product of `scene.environmentIntensity` AND `material.envMapIntensity` -- even if `scene.environment` is `null`.

>So, in that regard, would it be best to have scene.environmentIntensity apply only to scene.environment, and mesh.envMapIntensity apply only to mesh.envMap?

I think so, yes. It is also very clear.

If users want to have a different environment intensity for each mesh, then they can set `envMap` per mesh.

How does that sound?

**Comment by WestLangley** [comment]
>scene.environmentIntensity * material.envMapIntensity then?

After trying to implement this API, it seems weird to me...

The environment map = `scene.environment` OR `mesh.envMap` -- one or the other.

Yet, the intensity is the product of `scene.environmentIntensity` AND `material.envMapIntensity` -- even if `scene.environment` is `null`.

>So, in that regard, would it be best to have scene.environmentIntensity apply only to scene.environment, and mesh.envMapIntensity apply only to mesh.envMap?

I think so, yes. It is also very clear.

If users want to have a different environment intensity for each mesh, then they can set `envMap` per mesh.

How does that sound?

**Comment by WestLangley** [comment]
`

**Comment by viridia** [comment]
I can work with either of those approaches.

My primary use case here is day / night cycling in the game world. Currently I have the directional light (sunlight) as well as the ambient lights attenuated and tinted based on the time of day. However, since the global environment map never changes, things can look odd at night, depending on what type of shader is being used. Being able to set the environment intensity globally would fix this issue.

I don't use per-object environment maps at all, so I would only ever adjust the global intensity - it's not feasible to adjust the individual materials since there are so many in different places.

However, I'm not just thinking about my use case, but what other users might want. The current behavior is that the environment intensity setting on the material affects both global and per-mesh environment maps. The question is, how many existing users currently depend on the ability to be able to set the environment map globally, but be able to adjust the intensity locally?

One idea - a variation on what was proposed above - would be to have the global environment map intensity initially set to `undefined`. If the intensity is set to a number, then it behaves as you have suggested: local env maps are affected only by local intensity, and global env maps are only affected by global intensity. However, if the global intensity is undefined, then local intensity setting affects both.

In other words, the calculation for global intensity becomes:

```ts
const intensity = scene.envMapIntensity ?? material.envMapIntensity ?? 1;
```

This approach would be completely backwards compatible.

**Comment by iefreer** [comment]
blender has global environment intensity and rotation. the status of this ticket is closed, has the global scene environment enhancement been already implemented? thanks for having that.

**Comment by WestLangley** [comment]
`scene.environmentIntensity` has not been added due to conflicting viewpoints. See the suggestions in https://github.com/mrdoob/three.js/issues/24665#issuecomment-1250682600, https://github.com/mrdoob/three.js/issues/24665#issuecomment-1485694349 and https://github.com/mrdoob/three.js/issues/24665#issuecomment-1530190026.

**Comment by WestLangley** [comment]
>personally, i have hundreds of sandboxes where i use scene.environment + individual mesh.envMapIntensity.

Hmm... It appears that is true... `material.envMapIntensity` attenuates either `material.envMap` or `scene.environment` -- whichever environment happens to apply.

It works fine, guess. We definitely need to be able to attenuate `scene.environment`, and this does it.

Everybody OK with this?

**Comment by Archimagus** [comment]
and then in my SceneLighting component.
```ts
  useEffect(() => {
    scene.traverse((child) => {
      if (
        child instanceof Mesh &&
        child.material instanceof MeshStandardMaterial
      ) {
        child.material.envMapIntensity = globalLightIntensity;
      }
    });
  }, [globalLightIntensity, scene]);
```

Thats a lot of effort and overhead to apply a number to globally affect what is already a global value in 
```ts
    scene.environment = bgTexture;
```
